src/Maestro/DependencyUpdater/DependencyUpdater.cs

```diff
@@ -203,6 +207,63 @@ select new
             }
         }
 
+        /// <summary>
+        ///     Update Longest Build Path table once a day for each channel at midnight
+        /// </summary>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        [CronSchedule("0 0 0 1/1 * ? *", TimeZones.PST)]
+        public async Task UpdateLongestBuildPathAsync(CancellationToken cancellationToken)
+        {
+            IQueryable<Channel> query = Context.Channels;
+
+            List<Channel> channels = query.AsEnumerable().Select(c => new Channel() { Id = c.Id, Name = c.Name }).ToList();
+
+            // Get the flow graph
+            var barOnlyRemote = await RemoteFactory.GetBarOnlyRemoteAsync(Logger);
+
+            List<Microsoft.DotNet.Maestro.Client.Models.DefaultChannel> defaultChannels = (await barOnlyRemote.GetDefaultChannelsAsync()).ToList();
+            List<Microsoft.DotNet.Maestro.Client.Models.Subscription> subscriptions = (await barOnlyRemote.GetSubscriptionsAsync()).ToList();
+
+            IEnumerable<string> frequencies = 
+                new string[] { "everyWeek", "twiceDaily", "everyDay", "everyBuild", "none", };
+
+            foreach (var channel in channels)
+            {
+                // Build, then prune out what we don't want to see if the user specified
+                // channels.
+                DependencyFlowGraph flowGraph = await DependencyFlowGraph.BuildAsync(defaultChannels, subscriptions, barOnlyRemote, 30);
+
+                flowGraph.PruneGraph(
+                    node => DependencyFlowGraph.IsInterestingNode(channel.Name, node), 
+                    edge => DependencyFlowGraph.IsInterestingEdge(edge, false, frequencies));
+
+                if (flowGraph.Nodes.Count > 0)
+                {
+
+                    flowGraph.MarkBackEdges();
+                    flowGraph.CalculateLongestBuildPaths();
+                    flowGraph.MarkLongestBuildPath();
+
+                    // Get the nodes on the longest path and order them by path time so that the
+                    // contributing repos are in the right order
+                    IEnumerable<DependencyFlowNode> longestBuildPathNodes = flowGraph.Nodes.Where(n => n.OnLongestBuildPath).OrderByDescending(n => n.BestCasePathTime);
+
+                    LongestBuildPath lbp = new LongestBuildPath()
+                    {
+                        Channel = channel,
+                        BestCaseTimeInMinutes = longestBuildPathNodes.Max(n => n.BestCasePathTime),
+                        WorstCaseTimeInMinutes = longestBuildPathNodes.Max(n => n.WorstCasePathTime),
+                        ContributingRepositories = String.Join(';', longestBuildPathNodes.Select(n => $"{n.Repository}@{n.Branch}").ToArray()),
+                        EndDate = DateTimeOffset.UtcNow,
+                    };
+
+                    await Context.LongestBuildPaths.AddAsync(lbp);
```

Comment:
Should we check if we already built this build and not build it again?  Seems like we'll be dumping the same data in here over and over for the most part.  Or is that not an option because we don't have a way to identify "a build" of the product?

Reply:
If we can't, maybe just running once a day is better, since I expect most places will build at least once a day.

---
In reply to: [384089494](https://github.com/dotnet/arcade-services/pull/990#discussion_r384089494) [](ancestors = 384089494)

Reply:
This isn't being run per build. It's just a once per day calculation of the previous 30 days times at each repository. So it's the graph that we display on barviz, stored in history.